### PR TITLE
DevOps - Update Query node deployment for Giza

### PR DESCRIPTION
**Pull request by ahhda**
* Ability to use a pre-built docker 
  - Added a `useLocalRepo` config which uses the local image if true. If false will try to pull from the docker hub.
  - Added an `appsImage` config, which specifies the name (& tag) of the image to use
  - Added a dummy Dockerfile to build from a pre-built image as currently, there is no option to push an existing image to AWS ECR without building.
* Separate Indexer and Processor modules
* Reusable Database module added to `pulumi-common`
* Remove old migration jobs
* Update Caddy config not to require `/` (forward slashes)
* Add port names to `query-node` service to help with separate indexer deployment

<details>
<summary>Manual Testing</summary>

* Processor GraphQL https://34.197.252.42.nip.io/server/graphql
Screenshot 2021-11-02 at 1 09 24 PM

* Indexer GraphQL https://34.197.252.42.nip.io/indexer/graphql
Screenshot 2021-11-02 at 1 10 13 PM

</details>

### CODE REVIEW

File: devops/infrastructure/node-network/index.ts [hunk 1]
```diff
@@ -171,7 +171,9 @@ const chainDataPrepareJob = new k8s.batch.v1.Job(
               name: 'raw-chain-spec',
               image: nodeImage,
               command: ['/bin/sh', '-c'],
-              args: [`/joystream/node build-spec --chain ${chainDataPath}/chainspec.json --raw > ${chainSpecPath}`],
+              args: [
+                `sleep 15 && /joystream/node build-spec --chain ${chainDataPath}/chainspec.json --raw > ${chainSpecPath}`,
```
Comment (mnaamani) [inline devops/infrastructure/node-network/index.ts]: Why do we need to `sleep 15` ?

File: devops/infrastructure/query-node/index.ts [hunk 2]
```diff
@@ -56,6 +56,7 @@ if (isMinikube) {
     dockerfile: '../../../apps.Dockerfile',
     context: '../../../',
   })
+  // joystreamAppsImage = pulumi.interpolate`joystream/apps`
```
Comment (mnaamani) [inline devops/infrastructure/query-node/index.ts]: best not to have commented code, or at least add code comment why it is commented out.

File: devops/infrastructure/query-node/index.ts [hunk 3]
```diff
@@ -195,90 +196,8 @@ const migrationJob = new k8s.batch.v1.Job(
...
-    { path: membersFilePath, name: 'members.json' },
-    { path: workersFilePath, name: 'workers.json' },
-  ],
-  policy: s3Helpers.publicReadPolicy,
-})
-
-const membersUrl = dataBucket.getUrlForFile('members.json')
-const workersUrl = dataBucket.getUrlForFile('workers.json')
-
 const dataPath = '/joystream/query-node/mappings/bootstrap/data'
```
Comment (mnaamani) [inline devops/infrastructure/query-node/index.ts]: `dataPath` this variable is not used anymore and can be removed

File: devops/infrastructure/query-node/README.md [hunk 4]
```diff
@@ -71,6 +71,11 @@ After cloning this repo, from this working directory, run these commands:
 
    The GraphQl server is accessible at `https://<ENDPOINT>/server/graphql` and indexer at `https://<ENDPOINT>/indexer/graphql`
 
+1. If you are using Minikube, run `minikube service node-network -n $(pulumi stack output namespaceName)`
+
+   This will setup a proxy for your `query-node` service, which can then be accessed at
+   the URL given in the output
+
```
Comment (mnaamani) [inline devops/infrastructure/query-node/README.md]: instructions about configuring `--plaintext workersFilePath=<PATH> --plaintext membersFilePath=<PATH>` should also be removed earlier.

File: devops/infrastructure/query-node/index.ts [hunk 5]
```diff
@@ -295,11 +214,13 @@ const existingIndexer = config.get('indexerURL')
 if (!existingIndexer) {
   indexerContainer.push({
     name: 'indexer',
-    image: 'joystream/hydra-indexer:2.1.0-beta.9',
+    image: 'joystream/hydra-indexer:3.0.0',
```
Comment (mnaamani) [inline devops/infrastructure/query-node/index.ts]: The hydra-indexer-gateway and redis containers should be "grouped" with the indexer.

**Review by mnaamani — CHANGES_REQUESTED** [review]
Managed to deploy successfully. Left some suggestion for improvement.
One last request would be to be able to deploy only the "indexer" without a processor (just like we can deploy the "processor" without an indexer)

This will make it easier in production to re-deploy processors while keeping indexer untouched.

File: devops/kubernetes/query-node/index.ts [hunk 6]
```diff
@@ -4,31 +4,41 @@ import * as docker from '@pulumi/docker'
...
-  }).baseImageName
-  // joystreamAppsImage = pulumi.interpolate`joystream/apps`
+  if (useLocalRepo) {
+    // Use already existing image in minikube environment
+    joystreamAppsImage = pulumi.interpolate`${appsImage}`
+  } else {
+    // Access image from docker hub
+    joystreamAppsImage = new docker.RemoteImage('apps', {
+      name: appsImage!,
+    }).repoDigest
```
Comment (mnaamani) [inline devops/kubernetes/query-node/index.ts]: Despite setting both isMinikube and useLocalRepo to true in the config, code path is reaching this line and I'm getting error when bringing up the stack:

```
Previewing update (query-node):
     Type                 Name                   Plan       Info
 +   pulumi:pulumi:Stack  query-node-query-node  create     1 error
 
Diagnostics:
  pulumi:pulumi:Stack (query-node-query-node):
    error: Running program '/Users/mokhtar/joystream/joystream/devops/kubernetes/query-node' failed with an unhandled exception:
    TSError: ⨯ Unable to compile TypeScript:
    index.ts(40,8): error TS2339: Property 'repoDigest' does not exist on type 'RemoteImage'.
```

File: devops/kubernetes/query-node/index.ts [hunk 7]
```diff
@@ -4,31 +4,41 @@ import * as docker from '@pulumi/docker'
...
 
 require('dotenv').config()
 
 const config = new pulumi.Config()
 const awsConfig = new pulumi.Config('aws')
 const isMinikube = config.getBoolean('isMinikube')
+const externalIndexerUrl = config.get('externalIndexerUrl')
+const appsImage = config.get('appsImage') || `joystream/apps:latest`
+const skipProcessor = config.getBoolean('skipProcessor')
+const useLocalRepo = config.getBoolean('useLocalRepo') || false
```
Comment (mnaamani) [inline devops/kubernetes/query-node/index.ts]:
```suggestion
const useLocalRepo = config.getBoolean('useLocalRepo')
```

File: devops/kubernetes/query-node/index.ts (hunk 6 above)
Comment (mnaamani) [reply devops/kubernetes/query-node/index.ts]: Issue with useLocalRepo not being read correctly from config solved as suggested above

```suggestion
    }).name
```
Does that work?

Comment (ahhda) [reply devops/kubernetes/query-node/index.ts]: `repoDigest` was added in the most recent release of the `@pulumi/docker` package. But the name works as well. I'll update the code.

**Review by mnaamani — APPROVED** [review]

**Pull request merged**